# Exercise 4.5:- Data Consistency check 

Both the orders and Products dataset are checked for Data consistencies
1. Checked for missing values, duplicate records, mixed Data type.
2. Both orders and products datasets are exported back to the local Prepared Data Folder

#01 Importing Libraries

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

# Import the products Dataset and orders dataset after the wrangling step.

In [2]:
path = r'C:\Users\sorna\Desktop\Data Analysis\Achievement 4'

In [4]:
df_prods = df = pd.read_csv(os.path.join(path, '02.Data', 'Original Data', 'products.csv'), index_col = False)

In [3]:
df_ords = df = pd.read_csv(os.path.join(path, '02.Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_the_week  order_hour_of_day  days_since_prior_order  
count            3.421083e+06       3.421083e+06            3.214874e+06  
mean             2.776219e+00       1.345202e+01            1.111484e+01  
std              2.046829e+00       4.226088e+00            9.206737e+00  
min              0.000000e+00       0.000000e+00            0.000000e+00  
25%              1.000000e+00       1.000000e+01            4.000000e+00  
50%              3.000000e+00       1.300000e+01            7.000000e+00  
75%              5.000000e+00       1.600000e+01            1.500000e+01  
max              6.000000e+00       2.300000e+01            3.000000e+01

# 01 Creating a test dataframe for learning mixed data types

In [10]:
#create a dataframe
df_test = pd.DataFrame()

In [11]:
# create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [14]:
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
       print (col)

mix


C:\Users\sorna\AppData\Local\Temp\ipykernel_26156\1412710303.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [16]:
df_test['mix'] = df_test['mix'].astype('str')

In [17]:
df_test['mix'].dtype

dtype('O')

# 02 FIND THE NUMBER OF MISSING VALUES IN EACH COLUMN IN THE DATAFRAME

In [39]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

# 03 CREATE A NEW DATAFRAME WITH THE RECORDS HAVING MISSING VALUES

In [40]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [41]:
df_nan.head()

product_id product_name  aisle_id  department_id  prices
33           34          NaN       121             14    12.2
68           69          NaN        26              7    11.8
115         116          NaN        93              3    10.8
261         262          NaN       110             13    12.1
525         525          NaN       109             11     1.2

In [42]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [43]:
df_prods.shape

(49693, 5)

In [44]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

# 04 CREATE A NEW DATAFRAME DROPPING THE RECORDS WITH THE MISSING VALUES

In [45]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [46]:
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

# 05 CHECK FOR DULICATE RECORDS

In [47]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [48]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# 06 MOVE UNIQUE RECORDS TO A NEW DATAFRAME

In [49]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [50]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

# 07 EXPORT THE CONSISTENCY CHECKED PRODUCT DATAFRAME TO THE PREPARED DATA FOLDER

In [33]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02.Data', 'Prepared data', 'product_checked.csv'))

In [34]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_the_week  order_hour_of_day  days_since_prior_order  
count            3.421083e+06       3.421083e+06            3.214874e+06  
mean             2.776219e+00       1.345202e+01            1.111484e+01  
std              2.046829e+00       4.226088e+00            9.206737e+00  
min              0.000000e+00       0.000000e+00            0.000000e+00  
25%              1.000000e+00       1.000000e+01            4.000000e+00  
50%              3.000000e+00       1.300000e+01            7.000000e+00  
75%              5.000000e+00       1.600000e+01            1.500000e+01  
max              6.000000e+00       2.300000e+01            3.000000e+01

In [4]:
df_ords.shape

(3421083, 7)

# TASK 
# 01 df.describe on df_ords dataframe

In [7]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_the_week  order_hour_of_day  days_since_prior_order  
count            3.421083e+06       3.421083e+06            3.214874e+06  
mean             2.776219e+00       1.345202e+01            1.111484e+01  
std              2.046829e+00       4.226088e+00            9.206737e+00  
min              0.000000e+00       0.000000e+00            0.000000e+00  
25%              1.000000e+00       1.000000e+01            4.000000e+00  
50%              3.000000e+00       1.300000e+01            7.000000e+00  
75%              5.000000e+00       1.600000e+01            1.500000e+01  
max              6.000000e+00       2.300000e+01            3.000000e+01

# The maximum order_number is 100. 
# The max and min value for orders_day_of_the_week is 0 and 6, this represents 0 for Monday and 6 for Sunday. In total of 7 days of the week.
# The count of the days_since_prior_order is lower compared to the other columns. There could be missing values in this column

# 02 CHECK FOR MIXED DATA TYPES 

In [5]:
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
       print (col)

C:\Users\sorna\AppData\Local\Temp\ipykernel_8276\2710567795.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\sorna\AppData\Local\Temp\ipykernel_8276\2710567795.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\sorna\AppData\Local\Temp\ipykernel_8276\2710567795.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\sorna\AppData\Local\Temp\ipykernel_8276\2710567795.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\sorna\AppData\Local\Tem

# Received an error, so changed applymap to just map

In [6]:
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
       print (col)

# There are no columns with mixed data type in the Dataframe df_ords

In [7]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_the_week      int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

# 03 CHECK FOR MISSING VALUES

In [8]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_the_week         0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

# 04 ADDRESSING THE MISSING VALUES

# There are 206209 missing values in the column days_since_prior_order. The missing values could be for the customers during their first order. Values cannot be imputed, cannot be removed. Maybe we can introduce a new flag for 1st order.

# Before addressing the missing value, we need to check if the records with the missing values are having value 1 in the column 'order_number'.

In [9]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [10]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_the_week  \
0                 0   2539329        1             1                       2   
11               11   2168274        2             1                       2   
26               26   1374495        3             1                       1   
39               39   3343014        4             1                       6   
45               45   2717275        5             1                       3   
...             ...       ...      ...           ...                     ...   
3420930     3420930    969311   206205             1                       4   
3420934     3420934   3189322   206206             1                       3   
3421002     3421002   2166133   206207             1                       6   
3421019     3421019   2227043   206208             1                       1   
3421069     3421069   3154581   206209             1                       3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [11]:
df_ords_first_order = df_ords_nan[df_ords_nan['order_number'] == 1]

In [12]:
df_ords_first_order

Unnamed: 0  order_id  user_id  order_number  orders_day_of_the_week  \
0                 0   2539329        1             1                       2   
11               11   2168274        2             1                       2   
26               26   1374495        3             1                       1   
39               39   3343014        4             1                       6   
45               45   2717275        5             1                       3   
...             ...       ...      ...           ...                     ...   
3420930     3420930    969311   206205             1                       4   
3420934     3420934   3189322   206206             1                       3   
3421002     3421002   2166133   206207             1                       6   
3421019     3421019   2227043   206208             1                       1   
3421069     3421069   3154581   206209             1                       3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

# This shows that all records with missing values in the column days_since_prior_order are customer's first order. So these records cannot be deleted. Hence we can introduce a new flag to the dataframe.

In [13]:
df_ords['First_order'] = df_ords['days_since_prior_order'].isna()

In [30]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_the_week  \
0                 0   2539329        1             1                       2   
1                 1   2398795        1             2                       3   
2                 2    473747        1             3                       3   
3                 3   2254736        1             4                       4   
4                 4    431534        1             5                       4   
...             ...       ...      ...           ...                     ...   
3421078     3421078   2266710   206209            10                       5   
3421079     3421079   1854736   206209            11                       4   
3421080     3421080    626363   206209            12                       1   
3421081     3421081   2977660   206209            13                       1   
3421082     3421082    272231   206209            14                       6   

         order_hour_of_day  days_since_prior_order  First_order  
0                        8                     NaN         True  
1                        7                    15.0        False  
2                       12                    21.0        False  
3                        7                    29.0        False  
4                       15                    28.0        False  
...                    ...                     ...          ...  
3421078                 18                    29.0        False  
3421079                 10                    30.0        False  
3421080                 12                    18.0        False  
3421081                 12                     7.0        False  
3421082                 14                    30.0        False  

[3421083 rows x 8 columns]

In [33]:
df_ords_first_order = df_ords[df_ords['order_number'] == 1]

In [34]:
df_ords_first_order

Unnamed: 0  order_id  user_id  order_number  orders_day_of_the_week  \
0                 0   2539329        1             1                       2   
11               11   2168274        2             1                       2   
26               26   1374495        3             1                       1   
39               39   3343014        4             1                       6   
45               45   2717275        5             1                       3   
...             ...       ...      ...           ...                     ...   
3420930     3420930    969311   206205             1                       4   
3420934     3420934   3189322   206206             1                       3   
3421002     3421002   2166133   206207             1                       6   
3421019     3421019   2227043   206208             1                       1   
3421069     3421069   3154581   206209             1                       3   

         order_hour_of_day  days_since_prior_order  First_order  
0                        8                     NaN         True  
11                      11                     NaN         True  
26                      14                     NaN         True  
39                      11                     NaN         True  
45                      12                     NaN         True  
...                    ...                     ...          ...  
3420930                 12                     NaN         True  
3420934                 18                     NaN         True  
3421002                 19                     NaN         True  
3421019                 15                     NaN         True  
3421069                 11                     NaN         True  

[206209 rows x 8 columns]

# All first orders are marked with the flag.

# 05 CHECK FOR DUPLICATE VALUES

In [35]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [36]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_the_week, order_hour_of_day, days_since_prior_order, First_order]
Index: []

# There are no duplicated records in the ords Dataframe

# 06 EXPORT THE CLEANED DATAFRAMES

In [37]:
df_ords.to_csv(os.path.join(path, '02.Data', 'Prepared Data', 'orders_cleaned.csv'))

In [51]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02.Data', 'Prepared data', 'product_cleaned.csv'))